### PR TITLE
Remove output used only by sizes

### LINKED ISSUE #435: NVFuser Fusions produce intermediate Outputs were Legacy Fuser does not with `requires_grad=True`
## 🐛 Bug

The created kernel creates outputs for intermediate tensors with `requires_grad=True` option where the Legacy fuser does not.

This simple example fuses Add+Relu

Legacy Fuser's Kernel:

```
  } else {
    for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
         linearIndex < totalElements;
         linearIndex += gridDim.x * blockDim.x) {
      // Convert `linearIndex` into an offset of tensor:
      IndexType t0_offset = 0;
      IndexType t0_linearIndex = linearIndex;
      
      //printf("tensor t0 sizes[0] = %d, strides[0] = %d\n", t0.sizes[0],t0.strides[0]);
      size_t t0_dimIndex0 = t0_linearIndex ;
      t0_offset += t0_dimIndex0 ;
      IndexType t1_offset = 0;
      IndexType t1_linearIndex = linearIndex;
      
      //printf("tensor t1 sizes[0] = %d, strides[0] = %d\n", t1.sizes[0],t1.strides[0]);
      size_t t1_dimIndex0 = t1_linearIndex ;
      t1_offset += t1_dimIndex0 ;
      IndexType t2_offset = 0;
      IndexType t2_linearIndex = linearIndex;
      
      //printf("tensor t2 sizes[0] = %d, strides[0] = %d\n", t2.sizes[0],t2.strides[0]);
      size_t t2_dimIndex0 = t2_linearIndex ;
      t2_offset += t2_dimIndex0 ;
      
      // calculate the results
      float n0 = __half2float(t0.data[t0_offset]);
      float n1 = __half2float(t1.data[t1_offset]);
      int64_t n2 = 1;
      float n3 = n0 + ((float) n2)*n1;
      float n4 = n3 < 0 ? 0.f : n3 ;
      t2.data[t2_offset] = __float2half(n4);
      
    }
  }
}

graph(%x.1 : Tensor,
      %y.1 : Tensor):
  %15 : int = prim::BailoutTemplate_0()
  %13 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = prim::BailOut[index=0](%15, %y.1, %x.1)
  %14 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = prim::BailOut[index=1](%15, %x.1, %13)
  %17 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = prim::DifferentiableGraph_1(%14, %13)
  return (%17)
with prim::BailoutTemplate_0 = graph(%x.1 : Tensor,
      %y.1 : Tensor):
  %2 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = prim::BailOut[index=0](%y.1, %x.1)
  %3 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = prim::BailOut[index=1](%x.1, %2)
  %4 : int = prim::Constant[value=1]()
  %z.1 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = aten::add(%3, %2, %4) # add_relu.py:17:8
  %6 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = aten::relu(%z.1) # add_relu.py:18:11
  return (%6)
with prim::DifferentiableGraph_1 = graph(%2 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0),
      %3 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0)):
  %56 : int[]? = prim::Constant()
  %58 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = prim::FusionGroup_0(%2, %3)
  return (%58, %56, %56)
with prim::FusionGroup_0 = graph(%2 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0),
      %3 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0)):
  %4 : int = prim::Constant[value=1]()
  %z.2 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = aten::add(%2, %3, %4) # add_relu.py:17:8
  %1 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=1, device=cuda:0) = aten::relu(%z.2) # add_relu.py:18:11
  return (%1)

```

NVFuser's Kernel:

```
__global__ void kernel1(Tensor<__half, 4> T0, Tensor<__half, 4> T2, Tensor<__half, 4> T6, Tensor<__half, 4> T7) {
  float T1[1];
  float T3[1];
  float T4[1];
  float T5[1];
  if (((((((blockIdx.x * 1) + (1 - 1)) * 128) + threadIdx.x) < (((T0.size[0] * T0.size[1]) * T0.size[2]) * T0.size[3])) && (((((blockIdx.x * 1) + (1 - 1)) * 128) + threadIdx.x) < (((T0.size[0] * T0.size[1]) * T0.size[2]) * T0.size[3])))) {
    for(size_t i9 = 0; i9 < 1; ++i9) {
      T1[i9]
         = __half2float(T0[((((blockIdx.x * 1) + i9) * 128) + threadIdx.x)]);
      T3[i9]
         = __half2float(T2[((((blockIdx.x * 1) + i9) * 128) + threadIdx.x)]);
      T4[i9]
        = T1[i9]
        + T3[i9];
      T5[i9]
         = relu(T4[i9]);
      T6[((((blockIdx.x * 1) + i9) * 128) + threadIdx.x)]
         = __float2half(T5[i9]);
      T7[((((blockIdx.x * 1) + i9) * 128) + threadIdx.x)]
         = __float2half(T4[i9]);
    }
  } else {
    for(size_t i9 = 0; i9 < 1; ++i9) {
      if ((((((blockIdx.x * 1) + i9) * 128) + threadIdx.x) < (((T0.size[0] * T0.size[1]) * T0.size[2]) * T0.size[3]))) {
        T1[i9]
           = __half2float(T0[((((blockIdx.x * 1) + i9) * 128) + threadIdx.x)]);
      }
      if ((((((blockIdx.x * 1) + i9) * 128) + threadIdx.x) < (((T2.size[0] * T2.size[1]) * T2.size[2]) * T2.size[3]))) {
        T3[i9]
           = __half2float(T2[((((blockIdx.x * 1) + i9) * 128) + threadIdx.x)]);
      }
      if ((((((blockIdx.x * 1) + i9) * 128) + threadIdx.x) < (((T0.size[0] * T0.size[1]) * T0.size[2]) * T0.size[3]))) {
        T4[i9]
          = T1[i9]
          + T3[i9];
      }
      if ((((((blockIdx.x * 1) + i9) * 128) + threadIdx.x) < (((T0.size[0] * T0.size[1]) * T0.size[2]) * T0.size[3]))) {
        T5[i9]
           = relu(T4[i9]);
      }
      if ((((((blockIdx.x * 1) + i9) * 128) + threadIdx.x) < (((T0.size[0] * T0.size[1]) * T0.size[2]) * T0.size[3]))) {
        T6[((((blockIdx.x * 1) + i9) * 128) + threadIdx.x)]
           = __float2half(T5[i9]);
      }
      if ((((((blockIdx.x * 1) + i9) * 128) + threadIdx.x) < (((T0.size[0] * T0.size[1]) * T0.size[2]) * T0.size[3]))) {
        T7[((((blockIdx.x * 1) + i9) * 128) + threadIdx.x)]
           = __float2half(T4[i9]);
      }
    }
  }
}
}

======================================

graph(%x.1 : Tensor,
      %y.1 : Tensor):
  %10 : Tensor = prim::DifferentiableGraph_0(%y.1, %x.1)
  return (%10)
with prim::DifferentiableGraph_0 = graph(%10 : Tensor,
      %12 : Tensor):
  %50 : int[] = aten::size(%12) # <string>:3:44
  %53 : int[] = aten::size(%10) # <string>:3:93
  %66 : bool = prim::CudaFusionGuard[types=[Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0), Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0)]](%12, %10)
  %67 : Tensor, %68 : Tensor = prim::If(%66)
    block0():
      %64 : Tensor, %z.3 : Tensor = prim::CudaFusionGroup_0[cache_id=0](%12, %10)
      -> (%64, %z.3)
    block1():
      %74 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %75 : (Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0), Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0)) = prim::CallFunction(%74, %12, %10)
      %76 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0), %77 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0) = prim::TupleUnpack(%75)
      -> (%76, %77)
  %51 : int[] = aten::size(%68) # <string>:3:55
  %52 : int[]? = aten::_size_if_not_equal(%50, %51) # <string>:3:19
  %55 : int[]? = aten::_size_if_not_equal(%53, %51) # <string>:3:68
  return (%67, %52, %55, %67)
with prim::CudaFusionGroup_0 = graph(%2 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0),
      %3 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0)):
  %4 : int = prim::Constant[value=1]()
  %z.2 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0) = aten::add(%2, %3, %4) # add_relu.py:17:8
  %1 : Half(6:15564800, 200:77824, 304:256, 256:1, requires_grad=0, device=cuda:0) = aten::relu(%z.2) # add_relu.py:18:11
  return (%1, %z.2)

```

## To Reproduce

Toggle True = legacy and False = Nvfuser

```
import torch

## EDIT THIS CONDITIONAL
if True :
    torch._C._jit_set_profiling_executor(True)
    torch._C._jit_set_profiling_mode(True)
else :
    torch._C._jit_set_nvfuser_enabled(True)
    torch._C._jit_set_profiling_executor(True)
    torch._C._jit_set_profiling_mode(True)
    torch._C._jit_override_can_fuse_on_cpu(False)
    torch._C._jit_override_can_fuse_on_gpu(False)
    torch._C._jit_set_bailout_depth(20)


@torch.jit.script
def add_relu(x, y) :
    z = x + y
    return torch.relu(z)

input1 = torch.randn(6, 200, 304, 256, device='cuda', dtype=torch.float16, requires_grad=True)
input2 = torch.randn(6, 200, 304, 256, device='cuda', dtype=torch.float16, requires_grad=True)

for idx in range(0, 5) :
    if idx == 1:
        print(add_relu.graph_for(input1, input2))
    add_relu(input1, input2)
```

Command Line:

```
PYTORCH_CUDA_FUSER_DEBUG=1 PYTORCH_FUSION_DEBUG=1 python add_relu.py
```

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #435 

1. Re-enabled the pass to remove outputs from fusion that is only used by `aten::size`;
2. Added size computation for reduction op via new operator `prim::ReductionSizes`;

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp [hunk 1]
```diff
@@ -738,6 +737,38 @@ struct CudaGraphFuser {
...
+        // blocked by profiling executor not capable of profiling scalar inputs.
+        TORCH_INTERNAL_ASSERT(
+            n->input(1)->node()->kind() == prim::Constant &&
+                n->input(2)->node()->kind() == prim::Constant,
+            "only supports reduction axes and keepdim being constant");
+
+        // hmmm, do I need to setInsertPoint...
+        Node* in1_const =
+            graph->createClone(n->input(1)->node(), [](Value*) -> Value* {
+              throw std::runtime_error("unexpected input");
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/graph_fuser.cpp:754]: Is this throwing if `n->input(1)` has an input? Can you add a more sensible error, than unexpected input?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/graph_fuser.cpp:754]: I guess also if `n->input(1)` has a block, though I expect that's not the issue here.

**Review by csarofeen — APPROVED** [review]
I really have little idea what's going on in `buildShapeExpressions`, would be helpful if you could walk me through it, but I don't see anything that looks like a major error, so approving.

File: torch/csrc/jit/codegen/cuda/graph_fuser.cpp (hunk 1 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/graph_fuser.cpp:754]: Yes it does. in1_const is supposed to be a constant node, which was asserted earlier as well:
https://github.com/csarofeen/pytorch/pull/448/files/7747b0d3398d8e1667952d1c59ed60e0a689cb4e#diff-3f1e1ef27683f5e1c3d7d182d390823f7b3624c8b3a604d2b3a4567539cb4978R746-R749

I was `graph::createClone` for constant inputs, where you have to provide a mapping function from old inputs to new, hence the check and throw.

**Pull request merged**